cmd/kube-proxy/app/server.go

```diff
@@ -657,15 +657,33 @@ func detectNodeIPs(ctx context.Context, rawNodeIPs []net.IP, bindAddress string)
 		}
 	}
 
-	// If a bindAddress is passed, override the primary IP
-	bindIP := netutils.ParseIPSloppy(bindAddress)
-	if bindIP != nil && !bindIP.IsUnspecified() {
-		if netutils.IsIPv4(bindIP) {
-			primaryFamily = v1.IPv4Protocol
-		} else {
-			primaryFamily = v1.IPv6Protocol
+	// if nodeIPOverride is passed, override the IPs.
+	for i := range nodeIPOverride {
+		family := v1.IPv4Protocol
+		ip := netutils.ParseIPSloppy(nodeIPOverride[i])
+		if ip != nil && !ip.IsUnspecified() {
+			if !netutils.IsIPv4(ip) {
+				family = v1.IPv6Protocol
+			}
+			nodeIPs[family] = ip
+			// only override the primaryFamily for nodeIPOverride[0]
+			if i == 0 {
+				primaryFamily = family
+			}
+		}
+	}
+
+	// we should only return NodeIP for single IPFamily if IPFamilyPolicy is SingleStack.
+	if ipFamilyPolicy == kubeproxyconfig.IPFamilyPolicySingleStack {
+		var secondaryFamily v1.IPFamily
+		for family := range nodeIPs {
+			if family == primaryFamily {
+				continue
+			}
+			secondaryFamily = family
+			break
 		}
-		nodeIPs[primaryFamily] = bindIP
+		delete(nodeIPs, secondaryFamily)
 	}
```

Comment:
This is pretty confusing, and it's hard to be sure we got all the cases right...

I think it would be clearer if you figured out the primary IP family first, and then filled in nodeIPs separately after that.